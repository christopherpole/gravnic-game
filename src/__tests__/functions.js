import {
  MOVE_UP,
  ENTITIES,
  calulateNextGameState,
  changeGravityDirection,
  entitiesMatch,
  entitiesAreFading,
  levelIsComplete,
  isMatchableEntity,
} from '../index';

describe('calulateNextGameState()', () => {
  let gameState = [
    [{ staticEntity: { id: 1, entityId: ENTITIES.FLOOR }, movableEntity: null }],
    [{ staticEntity: { id: 2, entityId: ENTITIES.FLOOR }, movableEntity: null }],
    [
      {
        staticEntity: { id: 4, entityId: ENTITIES.FLOOR },
        movableEntity: { id: 3, entityId: ENTITIES.BLOCK },
      },
    ],
  ];

  it('Should return the next game state', () => {
    //  Move 1
    gameState = calulateNextGameState(gameState, MOVE_UP);
    expect(gameState).toMatchSnapshot();

    //  Move 2
    gameState = calulateNextGameState(gameState, MOVE_UP);
    expect(gameState).toMatchSnapshot();
  });

  it('Should return false if the entities cannot move any further', () => {
    gameState = calulateNextGameState(gameState, MOVE_UP);
    expect(gameState).toBe(false);
  });
});

describe('changeGravityDirection()', () => {
  const gameState = [
    [{ staticEntity: { id: 1, entityId: ENTITIES.FLOOR }, movableEntity: null }],
    [{ staticEntity: { id: 2, entityId: ENTITIES.FLOOR }, movableEntity: null }],
    [
      {
        staticEntity: { id: 4, entityId: ENTITIES.FLOOR },
        movableEntity: { id: 3, entityId: ENTITIES.BLOCK },
      },
    ],
  ];

  it('Should return an array of all game states from the move', () => {
    const gameStates = changeGravityDirection(gameState, MOVE_UP);

    expect(gameStates).toMatchSnapshot();
  });
});

describe('entitiesMatch()', () => {
  it('Matches two matching colored blocks together', () => {
    const matching = entitiesMatch(
      {
        entityId: ENTITIES.BLOCK,
        color: '#ff0000',
      },
      {
        entityId: ENTITIES.BLOCK,
        color: '#ff0000',
      },
    );

    expect(matching).toBe(true);
  });

  it("Doesn't match two non-matching colored blocks together", () => {
    const matching = entitiesMatch(
      {
        entityId: ENTITIES.BLOCK,
        color: '#00ff00',
      },
      {
        entityId: ENTITIES.BLOCK,
        color: '#ff0000',
      },
    );

    expect(matching).toBe(false);
  });
});

describe('entitiesAreFading()', () => {
  const gameState = [
    [{ staticEntity: { id: 1, entityId: ENTITIES.FLOOR }, movableEntity: null }],
    [{ staticEntity: { id: 2, entityId: ENTITIES.FLOOR }, movableEntity: null }],
    [
      {
        staticEntity: { id: 4, entityId: ENTITIES.FLOOR },
        movableEntity: { id: 3, entityId: ENTITIES.BLOCK },
      },
    ],
  ];

  it('Should return false if no entities are fading', () => {
    const fading = entitiesAreFading(gameState);

    expect(fading).toBe(false);
  });

  it('Should return true if entities are fading', () => {
    const fading = entitiesAreFading([
      ...gameState,
      [
        {
          staticEntity: { id: 5, entityId: ENTITIES.FLOOR },
          movableEntity: { id: 6, entityId: ENTITIES.BLOCK, fading: true },
        },
      ],
    ]);

    expect(fading).toBe(true);
  });
});

describe('levelIsComplete()', () => {
  const gameState = [
    [{ staticEntity: { id: 1, entityId: ENTITIES.FLOOR }, movableEntity: null }],
    [{ staticEntity: { id: 2, entityId: ENTITIES.FLOOR }, movableEntity: null }],
    [
      {
        staticEntity: { id: 4, entityId: ENTITIES.FLOOR },
        movableEntity: { id: 3, entityId: ENTITIES.BLOCK },
      },
    ],
  ];

  it('Should return false if there are matchable entities', () => {
    const complete = levelIsComplete(gameState);

    expect(complete).toBe(false);
  });

  it('Should return true if there are no matchable entities', () => {
    const complete = levelIsComplete(gameState.slice(0, 2));

    expect(complete).toBe(true);
  });
});

describe('isMatchableEntity()', () => {
  it('Should return "true" if the given entity ID is matchable', () => {
    const matchable = isMatchableEntity(ENTITIES.BLOCK);

    expect(matchable).toBe(true);
  });

  it('Should return "false" if the given entity ID is not matchable', () => {
    const matchable = isMatchableEntity(ENTITIES.GLASS);

    expect(matchable).toBe(false);
  });
});
