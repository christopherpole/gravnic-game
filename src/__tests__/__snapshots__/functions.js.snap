// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calulateNextGameState() Should return the next game state 1`] = `
Array [
  Array [
    Object {
      "movableEntity": null,
      "staticEntity": Object {
        "entityId": "FLOOR",
        "id": 1,
      },
    },
  ],
  Array [
    Object {
      "movableEntity": Object {
        "entityId": "BLOCK",
        "id": 3,
      },
      "staticEntity": Object {
        "entityId": "FLOOR",
        "id": 2,
      },
    },
  ],
  Array [
    Object {
      "movableEntity": null,
      "staticEntity": Object {
        "entityId": "FLOOR",
        "id": 4,
      },
    },
  ],
]
`;

exports[`calulateNextGameState() Should return the next game state 2`] = `
Array [
  Array [
    Object {
      "movableEntity": Object {
        "entityId": "BLOCK",
        "id": 3,
      },
      "staticEntity": Object {
        "entityId": "FLOOR",
        "id": 1,
      },
    },
  ],
  Array [
    Object {
      "movableEntity": null,
      "staticEntity": Object {
        "entityId": "FLOOR",
        "id": 2,
      },
    },
  ],
  Array [
    Object {
      "movableEntity": null,
      "staticEntity": Object {
        "entityId": "FLOOR",
        "id": 4,
      },
    },
  ],
]
`;

exports[`changeGravityDirection() Should return an array of all game states from the move 1`] = `
Array [
  Array [
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 1,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "BLOCK",
          "id": 3,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 4,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "movableEntity": Object {
          "entityId": "BLOCK",
          "id": 3,
        },
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 1,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 2,
        },
      },
    ],
    Array [
      Object {
        "movableEntity": null,
        "staticEntity": Object {
          "entityId": "FLOOR",
          "id": 4,
        },
      },
    ],
  ],
]
`;
